src/engine/Resources/Gif.ts

```diff
@@ -0,0 +1,567 @@
+  private _animation: ex.Animation = null;
+  private _transparentColor: ex.Color = null;
+
+  /**
+   * Populated once loading is complete
+   */
+  public images: HTMLImageElement;
+
+  /**
+   * @param path       Path to the image resource
+   * @param bustCache  Optionally load texture with cache busting
+   */
+  constructor(public path: string, public color: Color = Color.Magenta, public bustCache = true) {
+    super(path, 'arraybuffer', bustCache);
+    this._transparentColor = color;
+  }
+
+  /**
+   * Returns true if the Texture is completely loaded and is ready
+   * to be drawn.
+   */
+  public isLoaded(): boolean {
+    return this._isLoaded;
+  }
+
+  /**
+   * Begins loading the texture and returns a promise to be resolved on completion
+   */
+  public load(): Promise<Texture[]> {
+    var complete = new Promise<Texture[]>();
+    var loaded = super.load();
+    loaded.then(
+      () => {
+        this._stream = new Stream(this.getData());
+        this._gif = new ParseGif(this._stream, this._transparentColor);
+        const promises: Promise<HTMLImageElement>[] = [];
+        for (let imageIndex: number = 0; imageIndex < this._gif.images.length; imageIndex++) {
+          const texture = new Texture(this._gif.images[imageIndex].src, false);
+          this._texture.push(texture);
+          promises.push(texture.load());
+        }
+        Promise.join(promises).then(() => {
+          complete.resolve(this._texture);
+        });
+      },
+      () => {
+        complete.reject('Error loading texture.');
+      }
+    );
+    return complete;
+  }
+
+  public asSprite(id: number = 0): Sprite {
+    const sprite = this._texture[id].asSprite();
+    return sprite;
+  }
+
+  public asSpriteSheet(): SpriteSheet {
+    const spriteArray: Sprite[] = this._texture.map((texture) => {
+      return texture.asSprite();
+    });
+    return new SpriteSheet(spriteArray);
+  }
+
+  public asAnimation(engine: Engine, speed: number): ex.Animation {
+    const spriteSheet: SpriteSheet = this.asSpriteSheet();
+    this._animation = spriteSheet.getAnimationForAll(engine, speed);
+    return this._animation;
+  }
+}
+
+export interface Frame {
+  sentinel: number;
+  type: string;
+  leftPos: number;
+  topPos: number;
+  width: number;
+  height: number;
+  lctFlag: boolean;
+  interlaced: boolean;
+  sorted: boolean;
+  reserved: boolean[];
+  lctSize: number;
+  lzwMinCodeSize: number;
+  pixels: number[];
+}
+
+var bitsToNum = (ba: any) => {
+  return ba.reduce(function(s: number, n: number) {
+    return s * 2 + n;
+  }, 0);
+};
+
+var byteToBitArr = (bite: any) => {
+  var a = [];
+  for (var i = 7; i >= 0; i--) {
+    a.push(!!(bite & (1 << i)));
+  }
+  return a;
+};
+
+export class Stream {
+  data: any = null;
+  len: number = 0;
+  position: number = 0;
+
+  constructor(dataArray: ArrayBuffer) {
+    this.data = new Uint8Array(dataArray);
+    this.len = this.data.byteLength;
+    console.log(this.len);
```

Comment:
Hey, I will remove them and turn them into an array for the tests. These numbers should always be the same, if not then the parse failed. It will be really useful as a test. I plan to do this tonight or tomorrow.